File: agent-mesh-py/python/agent_mesh/amesh/listen.py
"""``amesh listen`` — bind QUIC, announce on mDNS, print received envelopes.

Mirror of ``agent-mesh-cli/src/listen.rs``. Each accepted envelope is
verified inside the pyo3 binding (cert chain, payload CID, agent
signature) before it lands here; we just format the audit record.
"""

from __future__ import annotations

import asyncio
import json
import sys
from pathlib import Path
from typing import Any, Dict, Optional

from agent_mesh.core import AgentKey, AgentMetadata, UserKey
from agent_mesh.discovery import AnnounceConfig, Announcer
from agent_mesh.transport import Endpoint

from . import util


def _payload_view(payload: bytes) -> Dict[str, Any]:
    """UTF-8 if valid, else hex. Mirrors `listen.rs::PayloadView`."""
    try:
        text = payload.decode("utf-8")
    except UnicodeDecodeError:
        return {"encoding": "hex", "hex": payload.hex()}
    return {"encoding": "utf8", "text": text}


async def run(home: Path, duration: Optional[str]) -> int:
    """Bind, announce, accept envelopes until duration / Ctrl-C.

    Returns 0 on a clean shutdown. Returns 1 if the user key is
    missing or duration parses badly. Per-envelope handshake failures
    are logged to stderr but don't abort the listener — same posture
    as the Rust CLI.
    """
    key_path = home / "user.key"
    if not key_path.exists():
        print(
            f"error: load {key_path} — run `amesh keygen` first",
            file=sys.stderr,
        )
        return 1
    user = UserKey.load(str(key_path))
    host = util.current_hostname()

    meta = AgentMetadata(
        role="amesh-listen",
        host=host,
        capabilities=[],
        issued_at=util.now_rfc3339(),
    )
    agent = AgentKey.issue(user, meta)
    agent_fp = agent.fingerprint()
    user_fp = user.fingerprint()
    cert = agent.cert()

    # Bind iroh first so we know the port to announce.
    endpoint = await Endpoint.bind(agent, 0)
    port = await endpoint.local_port()

    config = AnnounceConfig(
        agent_fp=agent_fp,
        user_fp=user_fp,
        capabilities=[],
        role="amesh-listen",
        host=host,
        port=port,
        agent_pubkey=bytes(agent.public_bytes()),
    )
    announcer = Announcer.start(config)

    print(f"listening on udp/{port}")
    print(f"  agent_fp={agent_fp.hex()}")
    print(f"  user_fp ={user_fp.hex()}")
    print(f"  host    ={host}")

    stop_deadline = None
    if duration is not None:
        try:
            dur_s = util.parse_duration(duration)
        except ValueError as e:
            announcer.stop()
            await endpoint.close()
            print(f"error: {e}", file=sys.stderr)
            return 1
        print(f"  duration={dur_s}s")
        stop_deadline = asyncio.get_event_loop().time() + dur_s
    else:
        print("  ctrl-c to stop")

    try:
        while True:
            if stop_deadline is not None:
                now = asyncio.get_event_loop().time()
                remaining = stop_deadline - now
                if remaining <= 0:
                    break
                # accept_envelope itself bounds its wait by
                # timeout_ms — pass the remaining window so the call
                # returns promptly when the deadline elapses.
                timeout_ms = max(1, int(remaining * 1000))
            else:
                # Long poll window so we don't busy-loop; Ctrl-C
                # still interrupts the awaitable.
                timeout_ms = 60_000
            try:
                accepted = await endpoint.accept_envelope(cert, timeout_ms)
            except Exception as e:  # noqa: BLE001 — log per-conn failures, keep listening
                print(f"connection error: {e}", file=sys.stderr)
                continue
            if accepted is None:
                # Timeout fired with no incoming connection — loop
                # around (or fall through to the deadline check).
                continue
            env = accepted.envelope
            record = {
                "sender_agent_fp": accepted.peer_agent_fp.hex(),
                "sender_user_fp": accepted.peer_user_fp.hex(),
                "sequence": env.sequence,
                "payload": _payload_view(env.payload()),
            }
            print(json.dumps(record))
    finally:
        announcer.stop()
        await endpoint.close()
    return 0

File: agent-mesh-cli/src/listen.rs
//! `amesh listen` — bind a QUIC endpoint, announce it on mDNS, accept
//! incoming envelopes, print each one as JSON to stdout.
//!
//! This is `amesh announce` + transport listener in one. While
//! `amesh announce` is discovery-only (TXT carries no pubkey, peers
//! can find but not dial), `amesh listen` publishes the pubkey *and*
//! the bound iroh UDP port — making the agent fully reachable.
//!
//! Each accepted connection runs an independent handshake task; a
//! single listener can serve many peers concurrently.

use std::path::PathBuf;
use std::time::Duration;

use crate::util;
use agent_mesh_core::{AgentKey, AgentMetadata, CertChain, UserKey};
use agent_mesh_discovery::{AnnounceConfig, Announcer};
use agent_mesh_transport::{do_handshake, iroh_reexports::Incoming, recv_envelope, Endpoint};
use anyhow::{Context, Result};
use serde::Serialize;

/// JSON record printed per received envelope.
#[derive(Debug, Serialize)]
struct ReceivedRecord {
    sender_agent_fp: String,
    sender_user_fp: String,
    sequence: u64,
    /// UTF-8 view of payload if it's valid UTF-8; otherwise hex.
    payload: PayloadView,
}

#[derive(Debug, Serialize)]
#[serde(tag = "encoding", rename_all = "snake_case")]
enum PayloadView {
    Utf8 { text: String },
    Hex { hex: String },
}

impl PayloadView {
    fn of(bytes: &[u8]) -> Self {
        match std::str::from_utf8(bytes) {
            Ok(s) => Self::Utf8 { text: s.into() },
            Err(_) => Self::Hex {
                hex: hex::encode(bytes),
            },
        }
    }
}

/// Run the `listen` subcommand.
pub async fn run(home: PathBuf, duration: Option<String>) -> Result<()> {
    let key_path = home.join("user.key");
    let user = UserKey::load(&key_path)
        .with_context(|| format!("load {} — run `amesh keygen` first", key_path.display()))?;
    let host = util::current_hostname();

    let agent = AgentKey::issue(
        &user,
        AgentMetadata {
            role: "amesh-listen".into(),
            host: host.clone(),
            capabilities: vec![],
            issued_at: util::now_rfc3339(),
            expires_at: None,
        },
    );
    let agent_fp = agent.fingerprint();
    let agent_pubkey = agent.public_bytes();
    let user_fp = user.fingerprint();

    // Bind iroh first so we know the port to announce.
    let endpoint = Endpoint::bind(&agent, 0).await?;
    let port = endpoint.port();

    let _announcer = Announcer::start(AnnounceConfig {
        agent_fp,
        agent_pubkey: Some(agent_pubkey),
        user_fp,
        capabilities: vec![],
        role: "amesh-listen".into(),
        host: host.clone(),
        port,
    })?;

    println!("listening on udp/{port}");
    println!("  agent_fp={}", agent_fp.hex());
    println!("  user_fp ={}", user_fp.hex());
    println!("  host    ={host}");
    let stop_deadline = if let Some(d) = duration {
        let dur = util::parse_duration(&d)?;
        println!("  duration={dur:?}");
        Some(tokio::time::Instant::now() + dur)
    } else {
        println!("  ctrl-c to stop");
        None
    };

    let cert = agent.cert().clone();

    loop {
        let accept_fut = endpoint.accept();
        let next = match stop_deadline {
            Some(deadline) => {
                tokio::select! {
                    biased;
                    _ = tokio::time::sleep_until(deadline) => break,
                    _ = tokio::signal::ctrl_c() => {
                        println!("ctrl-c received; stopping");
                        break;
                    },
                    next = accept_fut => next,
                }
            }
            None => {
                tokio::select! {
                    biased;
                    _ = tokio::signal::ctrl_c() => {
                        println!("ctrl-c received; stopping");
                        break;
                    },
                    next = accept_fut => next,
                }
            }
        };
        let Some(incoming) = next else {
            // Endpoint closed.
            break;
        };
        let cert = cert.clone();
        tokio::spawn(async move {
            if let Err(e) = handle_one(incoming, &cert).await {
                tracing::warn!(error = %e, "listener: connection error");
                eprintln!("connection error: {e}");
            }
        });
    }
    endpoint.close().await;
    Ok(())
}

async fn handle_one(incoming: Incoming, our_cert: &CertChain) -> Result<()> {
    let conn = incoming.await.context("await QUIC handshake")?;
    // First (and for now only) bidi stream per connection. Keep
    // accepting until the peer closes its send side.
    loop {
        let (mut send, mut recv) = match conn.accept_bi().await {
            Ok(streams) => streams,
            Err(e) => {
                tracing::debug!(error = %e, "accept_bi ended (peer closed)");
                return Ok(());
            }
        };
        // Use a deadline so a stalled handshake can't pin a stream.
        let peer_cert = match tokio::time::timeout(
            Duration::from_secs(5),
            do_handshake(our_cert, &mut send, &mut recv, false),
        )
        .await
        {
            Ok(Ok(cert)) => cert,
            Ok(Err(e)) => {
                eprintln!("rejected handshake: {e}");
                continue;
            }
            Err(_) => {
                eprintln!("handshake timed out after 5s");
                continue;
            }
        };
        // Now drain one envelope (Phase 2 ships one-shot envelopes
        // per stream; multi-message streams come later).
        match recv_envelope(&mut recv).await {
            Ok(env) => {
                let record = ReceivedRecord {
                    sender_agent_fp: env.sender_agent_fp().hex(),
                    sender_user_fp: env.sender_user_fp().hex(),
                    sequence: env.sequence,
                    payload: PayloadView::of(env.payload.as_ref()),
                };
                let line = serde_json::to_string(&record).expect("serialize record");
                println!("{line}");
                let _ = send.finish();
                // Reference peer_cert so it's not unused; the
                // tracing line below is the audit trail.
                tracing::debug!(peer = %peer_cert.agent_fingerprint().short(), "envelope received");
            }
            Err(e) => {
                eprintln!("envelope error: {e}");
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn payload_view_utf8_for_ascii() {
        match PayloadView::of(b"hello") {
            PayloadView::Utf8 { text } => assert_eq!(text, "hello"),
            other => panic!("expected utf8, got {other:?}"),
        }
    }

    #[test]
    fn payload_view_hex_for_invalid_utf8() {
        let bad = [0xffu8, 0xfe, 0xfd];
        match PayloadView::of(&bad) {
            PayloadView::Hex { hex } => assert_eq!(hex, "fffefd"),
            other => panic!("expected hex, got {other:?}"),
        }
    }
}
